src/components/PopupWithForm.js

```diff
@@ -1,68 +1,82 @@
+  constructor(popupSelector, handleSubmit, formValidator) {
     super({ popupSelector });
-    this._popup = document.querySelector(popupSelector);
-    this._handleFormSubmit = handleFormSubmit;
-    this._form = this._popup.querySelector("form");
-    this._inputs = Array.from(this._form.querySelectorAll("input"));
-    this._formData = {};
-    this._handleFormClose = handleFormClose;
+    this._handleSubmit = handleSubmit;
+    this._form = this.popupElement.querySelector("form");
+    this._submitButton = this._form.querySelector("button[type='submit']");
+    this._submitButtonText = this._submitButton.textContent;
+    this._formValidator = formValidator;
   }
 
-  _focusFirstInputField() {
-    const firstInput = this._inputs[0];
+  setEventListeners() {
+    super.setEventListeners();
+    this._form.addEventListener("submit", (evt) => {
+      evt.preventDefault();
+      this.renderLoading(true);
+
+      const submitPromise = this._handleSubmit(this._getInputValues());
+
+      if (!(submitPromise instanceof Promise)) {
+        console.error("The handleSubmit method did not return a promise.");
+        this.renderLoading(false);
+        return;
+      }
+
+      submitPromise
+        .then(() => this.close())
+        .catch((err) => console.error(`Error during form submission: ${err}`))
+        .finally(() => this.renderLoading(false));
+    });
+  }
+
+  resetForm() {
+    const form = this.popupElement.querySelector("form");
+    if (form) form.reset();
+  }
+
+  open(formData = {}) {
+    super.open();
+    this._populateForm(formData);
+    this._focusFirstInput();
+    this._formValidator.resetValidation();
+  }
+
+  _focusFirstInput() {
+    const firstInput = this._form.querySelector("input");
     if (firstInput) {
       firstInput.focus();
-      firstInput.select();
     }
   }
 
-  // have profile edit form initially display profile title and subtitle
-  setInputValues(data = {}) {
-    // const inputs = this._inputs;
-    this._inputs.forEach((input) => {
-      // put in the existing info or nothing
-      input.value = data[input.name] || "";
+  _populateForm(formData) {
+    Object.entries(formData).forEach(([name, value]) => {
+      const input = this._form.querySelector(`input[name="${name}"]`);
+      if (input) {
+        input.value = value;
+      }
     });
   }
 
-  // Open the popup and set initial values
-  open(data) {
-    this.setInputValues(data);
-    super.open();
-    this._focusFirstInputField();
-  }
-
   _getInputValues() {
-    const formValues = {};
-    // loop through every type of input field user can do
-    this._inputs.forEach((input) => {
-      // for each individual input field, use its name as a key in the
-      // inputValues object. then assign the user input to that key's value
-      formValues[input.name] = input.value;
+    const inputValues = {};
+    const inputs = this._form.querySelectorAll("input");
```

Comment:
`    const inputs = this._form.querySelectorAll("input");
`

You had `this._inputs` here